gateway/reverse_proxy_test.go

```diff
@@ -1034,6 +1034,148 @@ func TestGraphQL_InternalDataSource(t *testing.T) {
+		spec.APIID = "subgraph-bank-accounts"
+		spec.Proxy.TargetURL = TestHttpAny + bankAccountSubgraphPath
+		spec.Proxy.ListenPath = "/subgraph-bank-accounts"
+		spec.GraphQL = apidef.GraphQLConfig{
+			Enabled:       true,
+			ExecutionMode: apidef.GraphQLExecutionModeSubgraph,
+			Version:       apidef.GraphQLConfigVersion2,
+			Schema:        gqlSubgraphSchemaBankAccounts,
+			Subgraph: apidef.GraphQLSubgraphConfig{
+				SDL: gqlSubgraphSDLBankAccounts,
+			},
+		}
+	})[0]
+
+	t.Run("should batch requests", func(t *testing.T) {
+		supergraph := BuildAPI(func(spec *APISpec) {
+			spec.Proxy.ListenPath = "/batched-supergraph"
+			spec.APIID = "batched-supergraph"
+			spec.GraphQL = apidef.GraphQLConfig{
+				Enabled:       true,
+				Version:       apidef.GraphQLConfigVersion2,
+				ExecutionMode: apidef.GraphQLExecutionModeSupergraph,
+				Supergraph: apidef.GraphQLSupergraphConfig{
+					Subgraphs: []apidef.GraphQLSubgraphEntity{
+						{
+							APIID: "subgraph-accounts-modified",
+							URL:   "tyk://" + tykSubgraphAccounts.Name,
+							SDL:   gqlSubgraphSDLAccounts,
+						},
+						{
+							APIID: "subgraph-bank-accounts",
+							URL:   "tyk://" + tykSubgraphBankAccounts.Name,
+							SDL:   gqlSubgraphSDLBankAccounts,
+						},
+					},
+					MergedSDL: gqlMergedSupergraphSDL,
+				},
+				Schema: gqlMergedSupergraphSDL,
+			}
+		})[0]
+		g.Gw.LoadAPI(tykSubgraphAccounts, tykSubgraphBankAccounts, supergraph)
+		handlerCtx, cancel := context.WithCancel(context.Background())
+		g.AddDynamicHandler(bankAccountSubgraphPath, func(writer http.ResponseWriter, r *http.Request) {
+			select {
+			case <-handlerCtx.Done():
+				assert.Fail(t, "Called twice time")
+			default:
+			}
+			cancel()
+		})
+
+		q := graphql.Request{
+			Query: `query Query { allUsers { id username account { number } } }`,
+		}
+
+		_, _ = g.Run(t, []test.TestCase{
+			{
+				Data: q, Path: "/batched-supergraph",
+			},
+		}...)
+	})
+
+	t.Run("shouldn't batch requests", func(t *testing.T) {
+		supergraph := BuildAPI(func(spec *APISpec) {
+			spec.Proxy.ListenPath = "/unbatched-supergraph"
+			spec.APIID = "unbatched-supergraph"
+			spec.GraphQL = apidef.GraphQLConfig{
+				Enabled:       true,
+				Version:       apidef.GraphQLConfigVersion2,
+				ExecutionMode: apidef.GraphQLExecutionModeSupergraph,
+				Supergraph: apidef.GraphQLSupergraphConfig{
+					DisableQueryBatching: true,
+					Subgraphs: []apidef.GraphQLSubgraphEntity{
+						{
+							APIID: "subgraph-accounts-modified",
+							URL:   "tyk://" + tykSubgraphAccounts.Name,
+							SDL:   gqlSubgraphSDLAccounts,
+						},
+						{
+							APIID: "subgraph-bank-accounts",
+							URL:   "tyk://" + tykSubgraphBankAccounts.Name,
+							SDL:   gqlSubgraphSDLBankAccounts,
+						},
+					},
+					MergedSDL: gqlMergedSupergraphSDL,
+				},
+				Schema: gqlMergedSupergraphSDL,
+			}
+		})[0]
+		g.Gw.LoadAPI(tykSubgraphAccounts, tykSubgraphBankAccounts, supergraph)
+		timesHit := 0
+		lock := sync.Mutex{}
```

Comment:
A possible simplification for this case: 

```go
		var counter int32
		g.AddDynamicHandler(bankAccountSubgraphPath, func(writer http.ResponseWriter, r *http.Request) {
			atomic.AddInt32(&counter, 1)
		})

		q := graphql.Request{
			Query: `query Query { allUsers { id username account { number } } }`,
		}

		_, _ = g.Run(t, []test.TestCase{
			{
				Data: q, Path: "/unbatched-supergraph",
			},
		}...)
		// g.Run returned. We are done with the dynamic handler.
		assert.Equal(t, int32(2), atomic.LoadInt32(&counter))
```

The atomic increase may seem to be overkill but I think it is safe.